File: liblison/src/render.cpp

#include <lison/render.hpp>

#include <algorithm>
#include <ranges>
#include <variant>

namespace lison
{
    namespace detail
    {
	struct CoordTransformer
	{
	    double scale;

	    CoordTransformer(const Image &im, double ppi, double s)
		: scale { ppi / im.unit_per_inch * s }
	    {
	    }

	    double transform(double x) const
	    {
		return x * scale;
	    }

	    Point transform(const Point &p) const
	    {
		return Point {
		    transform(p.x),
		    transform(p.y)
		};
	    }
	};
    }

    static void render_shape(const Shape &, cairo_t *, const Image &, const detail::CoordTransformer &);

    namespace detail
    {
	struct PatternSetter
	{
	    cairo_t * const cr;
	    const CoordTransformer &transformer;

	    void operator()(const Monochrome &pattern) const
	    {
		cairo_set_source_rgba(
		    cr,
		    pattern.color.red,
		    pattern.color.green,
		    pattern.color.blue,
		    pattern.color.alpha);
	    }

	    void operator()(const LinearGradient &pattern) const
	    {
		Point p1 = transformer.transform(pattern.point1);
		Point p2 = transformer.transform(pattern.point2);
		cairo_pattern_t *pat = cairo_pattern_create_linear(
		    p1.x, p1.y,
		    p2.x, p2.y);
		cairo_pattern_add_color_stop_rgba(
		    pat, 0,
		    pattern.color1.red,
		    pattern.color1.green,
		    pattern.color1.blue,
		    pattern.color1.alpha);
		cairo_pattern_add_color_stop_rgba(
		    pat, 1,
		    pattern.color2.red,
		    pattern.color2.green,
		    pattern.color2.blue,
		    pattern.color2.alpha);

		cairo_set_source(cr, pat);

		cairo_pattern_destroy(pat);
	    }

	    void operator()(const RadialGradient &pattern) const
	    {
		Point c1 = transformer.transform(pattern.center1);
		Point c2 = transformer.transform(pattern.center2);
		cairo_pattern_t *pat = cairo_pattern_create_radial(
		    c1.x, c1.y, transformer.transform(pattern.radius1),
		    c2.x, c2.y, transformer.transform(pattern.radius2));
		cairo_pattern_add_color_stop_rgba(
		    pat, 0,
		    pattern.color1.red,
		    pattern.color1.green,
		    pattern.color1.blue,
		    pattern.color1.alpha);
		cairo_pattern_add_color_stop_rgba(
		    pat, 1,
		    pattern.color2.red,
		    pattern.color2.green,
		    pattern.color2.blue,
		    pattern.color2.alpha);

		cairo_set_source(cr, pat);

		cairo_pattern_destroy(pat);
	    }
	};

	struct SegmentPutter
	{
	    cairo_t * const cr;
	    const CoordTransformer &transformer;

	    void operator()(const LineSegment &seg) const
	    {
		Point p2 = transformer.transform(seg.point2);

		cairo_line_to(cr, p2.x, p2.y);
	    }

	    void operator()(const QuadraticBezierSegment &seg) const
	    {
		Point qp1;
		cairo_get_current_point(cr, &qp1.x, &qp1.y);
		Point qp2 = transformer.transform(seg.point2);
		Point qp3 = transformer.transform(seg.point3);

		Point cp2 = Point {
		    qp1.x + 2 * (qp2.x - qp1.x) / 3.0,
		    qp1.y + 2 * (qp2.y - qp1.y) / 3.0,
		};
		Point cp3 = Point {
		    qp3.x + 2 * (qp2.x - qp3.x) / 3.0,
		    qp3.y + 2 * (qp2.y - qp3.y) / 3.0,
		};
		Point cp4 = qp3;

		cairo_curve_to(cr, cp2.x, cp2.y, cp3.x, cp3.y, cp4.x, cp4.y);
	    }

	    void operator()(const CubicBezierSegment &seg) const
	    {
		Point p2 = transformer.transform(seg.point2);
		Point p3 = transformer.transform(seg.point3);
		Point p4 = transformer.transform(seg.point4);

		cairo_curve_to(cr, p2.x, p2.y, p3.x, p3.y, p4.x, p4.y);
	    }
	};

	struct ShapeRenderer
	{
	    cairo_t * const cr;
	    const Image &image;
	    const CoordTransformer &transformer;

	    void set_pattern(const Pattern &pattern) const
	    {
		PatternSetter setter { cr, transformer };
		std::visit(setter, pattern);
	    }

	    void set_pen(const Pen &pen) const
	    {
		set_pattern(pen.pattern);
		cairo_set_line_width(cr, transformer.transform(pen.width));

		cairo_set_line_cap(
		    cr,
		    pen.cap == LineCap::butt ? CAIRO_LINE_CAP_BUTT
		    : pen.cap == LineCap::round ? CAIRO_LINE_CAP_ROUND
		    : CAIRO_LINE_CAP_SQUARE);

		cairo_set_line_join(
		    cr,
		    pen.join == LineJoin::miter ? CAIRO_LINE_JOIN_MITER
		    : pen.join == LineJoin::round ? CAIRO_LINE_JOIN_ROUND
		    : CAIRO_LINE_JOIN_BEVEL);
	    }

	    void set_brush(const Brush &brush) const
	    {
		set_pattern(brush.pattern);
	    }

	    void put_path(const CurveData &data, bool closed) const
	    {
		Point start = transformer.transform(data.start);
		cairo_move_to(cr, start.x, start.y);

		SegmentPutter putter { cr, transformer };

		for (const Segment &seg : data.segments)
		{
		    std::visit(putter, seg);
		}

		if (closed)
		{
		    cairo_close_path(cr);
		}
	    }

	    void operator()(const Group &group) const
	    {
		for (const Shape &shape : group.content)
		{
		    render_shape(shape, cr, image, transformer);
		}
	    }

	    void operator()(const Curve &curve) const
	    {
		put_path(curve.data, false);

		set_pen(image.pens.at(curve.pen));
		cairo_stroke(cr);
	    }

	    void operator()(const Region &region) const
	    {
		put_path(region.data.curves.at(0), true);

		for (std::size_t i = 1; i < region.data.curves.size(); ++i)
		{
		    cairo_new_sub_path(cr);
		    put_path(region.data.curves[i], true);
		}

		if (region.brush.has_value())
		{
		    set_brush(image.brushes.at(region.brush.value()));
		    cairo_fill_preserve(cr);
		}

		if (region.pen.has_value())
		{
		    set_pen(image.pens.at(region.pen.value()));
		    cairo_stroke(cr);
		}
		else
		{
		    cairo_new_path(cr);
		}		    
	    }
	};
    }

    static void render_shape(const Shape &shape, cairo_t *cr, const Image &im, const detail::CoordTransformer &trans)
    {
	detail::ShapeRenderer renderer { cr, im, trans };
	std::visit(renderer, shape);
    }

    void render(const Image &im, cairo_t *cr, double ppi, double scale)
    {
	detail::CoordTransformer trans { im, ppi, scale };

	cairo_set_operator(cr, CAIRO_OPERATOR_OVER);
	cairo_set_fill_rule(cr, CAIRO_FILL_RULE_EVEN_ODD);
	cairo_new_path(cr);

	for (const Shape &shape : im.shapes)
	{
	    render_shape(shape, cr, im, trans);
	}
    }
}

File: src/render.rs

use crate::image::*;

use cairo::{Context, Result};

struct Scaler {
    factor: f64
}

impl Scaler {
    fn new(image: &Image, ppi: f64, scale: f64) -> Scaler {
        Scaler {
            factor: ppi / image.unit_per_inch * scale
        }
    }

    fn scale(&self, value: f64) -> f64 {
        value * self.factor
    }
}

pub fn render(context: &Context, image: &Image, ppi: f64, scale: f64) -> Result<()> {
    let scaler = Scaler::new(image, ppi, scale);

    context.set_operator(cairo::Operator::Over);
    context.set_fill_rule(cairo::FillRule::EvenOdd);
    context.new_path();

    for shape in image.shapes.iter() {
        render_shape(context, shape, image, &scaler)?;
    }

    Ok(())
}

fn render_shape(context: &Context, shape: &Shape, image: &Image, scaler: &Scaler) -> Result<()> {
    match shape {
        Shape::Group(group) => render_group(context, group, image, scaler),
        Shape::Curve(curve) => render_curve(context, curve, image, scaler),
        Shape::Region(region) => render_region(context, region, image, scaler)
    }
}

fn render_group(context: &Context, group: &GroupShape, image: &Image, scaler: &Scaler) -> Result<()> {
    for child in group.content.iter() {
        render_shape(context, child, image, scaler)?;
    }

    Ok(())
}

fn set_pattern(context: &Context, pattern: &Pattern, scaler: &Scaler) -> Result<()> {
    match pattern {
        Pattern::Monochrome(pat) => {
            context.set_source_rgba(pat.color.red, pat.color.green, pat.color.blue, pat.color.alpha);
        },
        Pattern::LinearGradient(pat) => {
            let grad = cairo::LinearGradient::new(
                scaler.scale(pat.point_1.x),
                scaler.scale(pat.point_1.y),
                scaler.scale(pat.point_2.x),
                scaler.scale(pat.point_2.y)
            );
            grad.add_color_stop_rgba(
                0.0,
                pat.color_1.red,
                pat.color_1.green,
                pat.color_1.blue,
                pat.color_1.alpha
            );
            grad.add_color_stop_rgba(
                1.0,
                pat.color_2.red,
                pat.color_2.green,
                pat.color_2.blue,
                pat.color_2.alpha
            );
            context.set_source(grad)?;
        },
        Pattern::RadialGradient(pat) => {
            let grad = cairo::RadialGradient::new(
                scaler.scale(pat.center_1.x),
                scaler.scale(pat.center_1.y),
                scaler.scale(pat.radius_1),
                scaler.scale(pat.center_2.x),
                scaler.scale(pat.center_2.y),
                scaler.scale(pat.radius_2),
            );
            grad.add_color_stop_rgba(
                0.0,
                pat.color_1.red,
                pat.color_1.green,
                pat.color_1.blue,
                pat.color_1.alpha
            );
            grad.add_color_stop_rgba(
                1.0,
                pat.color_2.red,
                pat.color_2.green,
                pat.color_2.blue,
                pat.color_2.alpha
            );
            context.set_source(grad)?;
        }
    }

    Ok(())
}

fn translate_line_cap(cap: LineCap) -> cairo::LineCap {
    match cap {
        LineCap::Butt => cairo::LineCap::Butt,
        LineCap::Round => cairo::LineCap::Round,
        LineCap::Square => cairo::LineCap::Square
    }
}

fn translate_line_join(join: LineJoin) -> cairo::LineJoin {
    match join {
        LineJoin::Miter => cairo::LineJoin::Miter,
        LineJoin::Round => cairo::LineJoin::Round,
        LineJoin::Bevel => cairo::LineJoin::Bevel
    }
}

fn set_pen(context: &Context, pen: &Pen, scaler: &Scaler) -> Result<()> {
    set_pattern(context, &pen.pattern, scaler)?;
    context.set_line_width(scaler.scale(pen.width));
    context.set_line_cap(translate_line_cap(pen.cap));
    context.set_line_join(translate_line_join(pen.join));

    Ok(())
}

fn set_brush(context: &Context, brush: &Brush, scaler: &Scaler) -> Result<()> {
    set_pattern(context, &brush.pattern, scaler)
}

fn plot_curve_data(context: &Context, data: &CurveData, scaler: &Scaler, closed: bool) -> Result<()> {
    context.move_to(scaler.scale(data.start.x), scaler.scale(data.start.y));

    for seg in data.segments.iter() {
        match seg {
            Segment::Line(line) => {
                context.line_to(scaler.scale(line.point_2.x), scaler.scale(line.point_2.y));
            },
            Segment::QuadraticBezier(bezier) => {
                let (x1, y1) = context.current_point()?;
                let x2 = scaler.scale(bezier.point_2.x);
                let y2 = scaler.scale(bezier.point_2.y);
                let x3 = scaler.scale(bezier.point_3.x);
                let y3 = scaler.scale(bezier.point_3.y);
                context.curve_to(
                    1.0 / 3.0 * x1 + 2.0 / 3.0 * x2,
                    1.0 / 3.0 * y1 + 2.0 / 3.0 * y2,
                    1.0 / 3.0 * x3 + 2.0 / 3.0 * x2,
                    1.0 / 3.0 * y3 + 2.0 / 3.0 * y2,
                    x3,
                    y3
                );
            },
            Segment::CubicBezier(bezier) => {
                context.curve_to(
                    scaler.scale(bezier.point_2.x),
                    scaler.scale(bezier.point_2.y),
                    scaler.scale(bezier.point_3.x),
                    scaler.scale(bezier.point_3.y),
                    scaler.scale(bezier.point_4.x),
                    scaler.scale(bezier.point_4.y)
                );
            }
        }
    }

    if closed {
        context.close_path();
    }

    Ok(())
}

fn render_curve(context: &Context, curve: &CurveShape, image: &Image, scaler: &Scaler) -> Result<()> {
    plot_curve_data(context, &curve.data, scaler, false)?;

    if curve.pen >= image.pens.len() {
        panic!("invalid pen index {}, must be less than {}.", curve.pen, image.pens.len());
    }

    set_pen(context, &image.pens[curve.pen], scaler)?;
    context.stroke()
}

fn render_region(context: &Context, region: &RegionShape, image: &Image, scaler: &Scaler) -> Result<()> {
    if region.data.len() != 0 {
        plot_curve_data(context, &region.data[0], scaler, true)?;
    }

    for i in 1..region.data.len() {
        context.new_sub_path();
        plot_curve_data(context, &region.data[i], scaler, true)?;
    }

    if let Some(brush) = region.brush {
        if brush >= image.brushes.len() {
            panic!("invalid brush index {}, must be less than {}.", brush, image.brushes.len());
        }

        set_brush(context, &image.brushes[brush], scaler)?;
        context.fill_preserve()?;
    }

    if let Some(pen) = region.pen {
        if pen >= image.pens.len() {
            panic!("invalid pen index {}, must be less than {}.", pen, image.pens.len());
        }

        set_pen(context, &image.pens[pen], scaler)?;
        context.stroke()?;
    } else {
        context.new_path();
    }

    Ok(())
}
